core/src/main/java/io/questdb/cairo/vm/api/MemoryCR.java

```diff
@@ -0,0 +1,164 @@
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.vm.api;
+
+import io.questdb.cairo.CairoException;
+import io.questdb.cairo.TableUtils;
+import io.questdb.cairo.vm.Vm;
+import io.questdb.std.*;
+import io.questdb.std.str.AbstractCharSequence;
+import io.questdb.std.str.CharSink;
+
+public interface MemoryCR extends MemoryC, MemoryR {
+    default BinarySequence getBin(long offset, ByteSequenceView view) {
+        final long addr = addressOf(offset);
+        final long len = Unsafe.getUnsafe().getLong(addr);
+        if (len > -1) {
+            return view.of(addr + Long.BYTES, len);
+        }
+        return null;
+    }
+
+    default long getBinLen(long offset) {
+        return getLong(offset);
+    }
+
+    default boolean getBool(long offset) {
+        return getByte(offset) == 1;
+    }
+
+    default byte getByte(long offset) {
+        assert addressOf(offset + Byte.BYTES) > 0;
+        return Unsafe.getUnsafe().getByte(addressOf(offset));
+    }
+
+    default double getDouble(long offset) {
+        assert addressOf(offset + Double.BYTES) > 0;
+        return Unsafe.getUnsafe().getDouble(addressOf(offset));
+    }
+
+    default float getFloat(long offset) {
+        assert addressOf(offset + Float.BYTES) > 0;
+        return Unsafe.getUnsafe().getFloat(addressOf(offset));
+    }
+
+    default int getInt(long offset) {
+        assert addressOf(offset + Integer.BYTES) > 0;
+        return Unsafe.getUnsafe().getInt(addressOf(offset));
+    }
+
+    default long getLong(long offset) {
+        assert addressOf(offset + Long.BYTES) > 0;
+        return Unsafe.getUnsafe().getLong(addressOf(offset));
+    }
+
+    default long getPageSize() {
+        return size();
+    }
+
+    default short getShort(long offset) {
+        return Unsafe.getUnsafe().getShort(addressOf(offset));
+    }
+
+    default void getLong256(long offset, CharSink sink) {
+        final long addr = addressOf(offset + Long256.BYTES);
+        final long a, b, c, d;
+        a = Unsafe.getUnsafe().getLong(addr - Long.BYTES * 4);
+        b = Unsafe.getUnsafe().getLong(addr - Long.BYTES * 3);
+        c = Unsafe.getUnsafe().getLong(addr - Long.BYTES * 2);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).